JPA Map Storage with CRDB: UserSessionProviderTest failures

### Describe the bug

**org.keycloak.testsuite.model.UserSessionProviderTest** presents failures/errors with `CockroachDB`:

```
[ERROR]   UserSessionProviderTest.testOnUserRemoved:754 expected:<0> but was:<2>
[ERROR]   UserSessionProviderTest.testRemoveUserSessionsByUser:273 expected:<0> but was:<2>
```

- Required JPA stores to reproduce the failures **userSessions**

In addition to these failures, that can be reproduced by simply configuring userSessions to use JPA storage, there are other failures when other stores are also set to JPA:

```
[ERROR]   UserSessionProviderTest.testGetByClient:544->assertSessions:800 array lengths differed, expected.length=3 actual.length=0; arrays first differed at element [0]; expected:<0cfa44ca-3ca0-42a6-b839-34f9cf62bc7c> but was:<end of array>
[ERROR]   UserSessionProviderTest.testGetByClientPaginated:572->assertPaginatedSession:668 expected:<1> but was:<0>
[ERROR]   UserSessionProviderTest.testGetCountByClient:679 expected:<3> but was:<0>
[ERROR]   UserSessionProviderTest.testGetUserSessions:253->assertSessions:800 array lengths differed, expected.length=2 actual.length=0; arrays first differed at element [0]; expected:<2bbb2239-c68b-453f-a277-e1fed458fead> but was:<end of array>
[ERROR]   UserSessionProviderTest.testRemoveUserSessionsByExpired:385

[ERROR]   UserSessionProviderTest.testCreateClientSession » RunOnServer java.lang.NullPointerException
[ERROR]   UserSessionProviderTest.testCreateSessions » RunOnServer java.lang.NullPointerException
[ERROR]   UserSessionProviderTest.testOnClientRemoved » InternalServerError HTTP 500 Internal Server Error
[ERROR]   UserSessionProviderTest.testRestartSession » RunOnServer java.lang.NullPointerException
[ERROR]   UserSessionProviderTest.testUpdateClientSession » RunOnServer java.lang.NullPointerException
[ERROR]   UserSessionProviderTest.testUpdateClientSessionInSameTransaction » RunOnServer java.lang.NullPointerException
[ERROR]   UserSessionProviderTest.testUpdateClientSessionWithGetByClientId » RunOnServer java.lang.NullPointerException
[ERROR]   UserSessionProviderTest.testUpdateSession » RunOnServer java.lang.NullPointerException
[ERROR]   UserSessionProviderTest.testUpdateSessionInSameTransaction » RunOnServer java.lang.NullPointerException
```

- Required JPA stores to reproduce the failures **userSessions**, **realms**, and **clients**

The NPEs reported here resemble the one reported in #13211, in that hibernate is not able to locate entries that exist in the database (in this case particular user sessions). Also, failures seen in tests like `testOnUserRemoved` also have to do with the fact that hibernate doesn't seem to detect that session entries were actually removed from the database and is returning some sort of stale data.

Other tests, such as `testOnClientRemoved` are failing due to transaction issues:

```
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh): "sql txn" meta={id=4bfd0a61 key=/Table/3170/2/1/"OƍS\x91nAv\xb3Ϭ\xe9G\xe3\x18\xc9"/0 pri=0.01106863 epo=0 ts=1658256011.189340483,1 min=1658256011.182710187,0 seq=3} lock=true stat=PENDING rts=1658256011.182710187,0 wto=false gul=1658256011.682710187,0
  Hint: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html#retry_serializable
```


### Version

main

### Expected behavior

Tests should run without failures or errors



### Actual behavior

_No response_

### How to Reproduce?

1- Setup CockroachDB according to the instructions in https://github.com/keycloak/keycloak/discussions/9596
2- Run any of the failing tests with the correct parameters, enabling JPA storage for **userSessions**, **realms** and **clients**:

```
cd testsuite/integration-arquillian

mvn clean install -Pmap-storage -Dpageload.timeout=90000 -Dkeycloak.realm.map.storage.provider=jpa \
-Dkeycloak.client.map.storage.provider=jpa \
-Dkeycloak.userSession.map.storage.provider=jpa \
-Dkeycloak.map.storage.connectionsJpa.url=jdbc:postgresql://localhost:26257/keycloak \
-Dkeycloak.map.storage.connectionsJpa.user=keycloak -Dkeycloak.map.storage.connectionsJpa.password=pass \
-Dtest=UserSessionProviderTest
```

### Anything else?

_No response_

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderTest.java
@@ -107,10 +107,11 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         int started = Time.currentTime();
         RealmModel realm = session.realms().getRealmByName("test");
         UserSessionModel[] sessions = createSessions(session);
-
-        assertSession(session.sessions().getUserSession(realm, sessions[0].getId()), session.users().getUserByUsername(realm, "user1"), "127.0.0.1", started, started, "test-app", "third-party");
-        assertSession(session.sessions().getUserSession(realm, sessions[1].getId()), session.users().getUserByUsername(realm, "user1"), "127.0.0.2", started, started, "test-app");
-        assertSession(session.sessions().getUserSession(realm, sessions[2].getId()), session.users().getUserByUsername(realm, "user2"), "127.0.0.3", started, started, "test-app");
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            assertSession(kcSession.sessions().getUserSession(realm, sessions[0].getId()), session.users().getUserByUsername(realm, "user1"), "127.0.0.1", started, started, "test-app", "third-party");
+            assertSession(kcSession.sessions().getUserSession(realm, sessions[1].getId()), session.users().getUserByUsername(realm, "user1"), "127.0.0.2", started, started, "test-app");
+            assertSession(kcSession.sessions().getUserSession(realm, sessions[2].getId()), session.users().getUserByUsername(realm, "user2"), "127.0.0.3", started, started, "test-app");
+        });
     }
 
     @Test
@@ -119,9 +120,10 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         RealmModel realm = session.realms().getRealmByName("test");
         UserSessionModel[] sessions = createSessions(session);
         int lastRefresh = Time.currentTime();
-        session.sessions().getUserSession(realm, sessions[0].getId()).setLastSessionRefresh(lastRefresh);
-
-        assertEquals(lastRefresh, session.sessions().getUserSession(realm, sessions[0].getId()).getLastSessionRefresh());
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            kcSession.sessions().getUserSession(realm, sessions[0].getId()).setLastSessionRefresh(lastRefresh);
+            assertEquals(lastRefresh, kcSession.sessions().getUserSession(realm, sessions[0].getId()).getLastSessionRefresh());
+        });
     }
 
     @Test
@@ -130,8 +132,10 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         RealmModel realm = session.realms().getRealmByName("test");
         UserSessionModel[] sessions = createSessions(session);
         int lastRefresh = Time.currentTime();
-        session.sessions().getUserSession(realm, sessions[0].getId()).setLastSessionRefresh(lastRefresh);
-        assertEquals(lastRefresh, session.sessions().getUserSession(realm, sessions[0].getId()).getLastSessionRefresh());
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            kcSession.sessions().getUserSession(realm, sessions[0].getId()).setLastSessionRefresh(lastRefresh);
+            assertEquals(lastRefresh, kcSession.sessions().getUserSession(realm, sessions[0].getId()).getLastSessionRefresh());
+        });
     }
 
     @Test
@@ -143,14 +147,15 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
 
         Time.setOffset(100);
         try {
+            KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+                UserSessionModel userSession = kcSession.sessions().getUserSession(realm, sessions[0].getId());
+                assertSession(userSession, kcSession.users().getUserByUsername(realm, "user1"), "127.0.0.1", started, started, "test-app", "third-party");
 
-            UserSessionModel userSession = session.sessions().getUserSession(realm, sessions[0].getId());
-            assertSession(userSession, session.users().getUserByUsername(realm, "user1"), "127.0.0.1", started, started, "test-app", "third-party");
+                userSession.restartSession(realm, kcSession.users().getUserByUsername(realm, "user2"), "user2", "127.0.0.6", "form", true, null, null);
 
-            userSession.restartSession(realm, session.users().getUserByUsername(realm, "user2"), "user2", "127.0.0.6", "form", true, null, null);
-
-            userSession = session.sessions().getUserSession(realm, sessions[0].getId());
-            assertSession(userSession, session.users().getUserByUsername(realm, "user2"), "127.0.0.6", started + 100, started + 100);
+                userSession = kcSession.sessions().getUserSession(realm, sessions[0].getId());
+                assertSession(userSession, kcSession.users().getUserByUsername(realm, "user2"), "127.0.0.6", started + 100, started + 100);
+            });
         } finally {
             Time.setOffset(0);
         }
@@ -163,18 +168,20 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         RealmModel realm = session.realms().getRealmByName("test");
         UserSessionModel[] sessions = createSessions(session);
 
-        Map<String, AuthenticatedClientSessionModel> clientSessions = session.sessions().getUserSession(realm, sessions[0].getId()).getAuthenticatedClientSessions();
-        assertEquals(2, clientSessions.size());
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            Map<String, AuthenticatedClientSessionModel> clientSessions = kcSession.sessions().getUserSession(realm, sessions[0].getId()).getAuthenticatedClientSessions();
+            assertEquals(2, clientSessions.size());
 
-        String clientUUID = realm.getClientByClientId("test-app").getId();
+            String clientUUID = realm.getClientByClientId("test-app").getId();
 
-        AuthenticatedClientSessionModel session1 = clientSessions.get(clientUUID);
+            AuthenticatedClientSessionModel session1 = clientSessions.get(clientUUID);
 
-        assertNull(session1.getAction());
-        assertEquals(realm.getClientByClientId("test-app").getClientId(), session1.getClient().getClientId());
-        assertEquals(sessions[0].getId(), session1.getUserSession().getId());
-        assertEquals("http://redirect", session1.getRedirectUri());
-        assertEquals("state", session1.getNote(OIDCLoginProtocol.STATE_PARAM));
+            assertNull(session1.getAction());
+            assertEquals(realm.getClientByClientId("test-app").getClientId(), session1.getClient().getClientId());
+            assertEquals(sessions[0].getId(), session1.getUserSession().getId());
+            assertEquals("http://redirect", session1.getRedirectUri());
+            assertEquals("state", session1.getNote(OIDCLoginProtocol.STATE_PARAM));
+        });
     }
 
     @Test
@@ -187,18 +194,20 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         String userSessionId = sessions[0].getId();
         String clientUUID = realm.getClientByClientId("test-app").getId();
 
-        UserSessionModel userSession = session.sessions().getUserSession(realm, userSessionId);
-        AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessions().get(clientUUID);
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            UserSessionModel userSession = kcSession.sessions().getUserSession(realm, userSessionId);
+            AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessions().get(clientUUID);
 
-        int time = clientSession.getTimestamp();
-        assertNull(clientSession.getAction());
+            int time = clientSession.getTimestamp();
+            assertNull(clientSession.getAction());
 
-        clientSession.setAction(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name());
-        clientSession.setTimestamp(time + 10);
+            clientSession.setAction(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name());
+            clientSession.setTimestamp(time + 10);
 
-        AuthenticatedClientSessionModel updated = session.sessions().getUserSession(realm, userSessionId).getAuthenticatedClientSessions().get(clientUUID);
-        assertEquals(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name(), updated.getAction());
-        assertEquals(time + 10, updated.getTimestamp());
+            AuthenticatedClientSessionModel updated = kcSession.sessions().getUserSession(realm, userSessionId).getAuthenticatedClientSessions().get(clientUUID);
+            assertEquals(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name(), updated.getAction());
+            assertEquals(time + 10, updated.getTimestamp());
+        });
     }
 
     @Test
@@ -210,18 +219,20 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         String userSessionId = sessions[0].getId();
         String clientUUID = realm.getClientByClientId("test-app").getId();
 
-        UserSessionModel userSession = session.sessions().getUserSession(realm, userSessionId);
-        AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(clientUUID);
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            UserSessionModel userSession = kcSession.sessions().getUserSession(realm, userSessionId);
+            AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(clientUUID);
 
-        int time = clientSession.getTimestamp();
-        assertNull(clientSession.getAction());
+            int time = clientSession.getTimestamp();
+            assertNull(clientSession.getAction());
 
-        clientSession.setAction(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name());
-        clientSession.setTimestamp(time + 10);
+            clientSession.setAction(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name());
+            clientSession.setTimestamp(time + 10);
 
-        AuthenticatedClientSessionModel updated = session.sessions().getUserSession(realm, userSessionId).getAuthenticatedClientSessionByClient(clientUUID);
-        assertEquals(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name(), updated.getAction());
-        assertEquals(time + 10, updated.getTimestamp());
+            AuthenticatedClientSessionModel updated = kcSession.sessions().getUserSession(realm, userSessionId).getAuthenticatedClientSessionByClient(clientUUID);
+            assertEquals(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name(), updated.getAction());
+            assertEquals(time + 10, updated.getTimestamp());
+        });
     }
 
     @Test
@@ -233,15 +244,17 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         String userSessionId = sessions[0].getId();
         String clientUUID = realm.getClientByClientId("test-app").getId();
 
-        UserSessionModel userSession = session.sessions().getUserSession(realm, userSessionId);
-        AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(clientUUID);
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            UserSessionModel userSession = kcSession.sessions().getUserSession(realm, userSessionId);
+            AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(clientUUID);
 
-        clientSession.setAction(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name());
-        clientSession.setNote("foo", "bar");
+            clientSession.setAction(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name());
+            clientSession.setNote("foo", "bar");
 
-        AuthenticatedClientSessionModel updated = session.sessions().getUserSession(realm, userSessionId).getAuthenticatedClientSessionByClient(clientUUID);
-        assertEquals(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name(), updated.getAction());
-        assertEquals("bar", updated.getNote("foo"));
+            AuthenticatedClientSessionModel updated = kcSession.sessions().getUserSession(realm, userSessionId).getAuthenticatedClientSessionByClient(clientUUID);
+            assertEquals(AuthenticatedClientSessionModel.Action.LOGGED_OUT.name(), updated.getAction());
+            assertEquals("bar", updated.getNote("foo"));
+        });
     }
 
     @Test
@@ -250,10 +263,12 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         RealmModel realm = session.realms().getRealmByName("test");
         UserSessionModel[] sessions = createSessions(session);
 
-        assertSessions(session.sessions().getUserSessionsStream(realm, session.users().getUserByUsername(realm, "user1"))
-                .collect(Collectors.toList()), sessions[0], sessions[1]);
-        assertSessions(session.sessions().getUserSessionsStream(realm, session.users().getUserByUsername(realm, "user2"))
-                .collect(Collectors.toList()), sessions[2]);
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            assertSessions(kcSession.sessions().getUserSessionsStream(realm, session.users().getUserByUsername(realm, "user1"))
+                    .collect(Collectors.toList()), sessions[0], sessions[1]);
+            assertSessions(kcSession.sessions().getUserSessionsStream(realm, session.users().getUserByUsername(realm, "user2"))
+                    .collect(Collectors.toList()), sessions[2]);
+        });
     }
 
     @Test
@@ -262,25 +277,28 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         RealmModel realm = session.realms().getRealmByName("test");
         createSessions(session);
 
-        Map<String, Integer> clientSessionsKept = session.sessions().getUserSessionsStream(realm,
-                session.users().getUserByUsername(realm, "user2"))
-                .collect(Collectors.toMap(model -> model.getId(), model -> model.getAuthenticatedClientSessions().keySet().size()));
-
+        final Map<String, Integer> clientSessionsKept = new HashMap<>();
         KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession kcSession) -> {
+            clientSessionsKept.putAll(kcSession.sessions().getUserSessionsStream(realm,
+                            kcSession.users().getUserByUsername(realm, "user2"))
+                    .collect(Collectors.toMap(model -> model.getId(), model -> model.getAuthenticatedClientSessions().keySet().size())));
+
             kcSession.sessions().removeUserSessions(realm, kcSession.users().getUserByUsername(realm, "user1"));
         });
 
-        assertEquals(0, session.sessions().getUserSessionsStream(realm, session.users().getUserByUsername(realm, "user1"))
-                .count());
-        List<UserSessionModel> userSessions = session.sessions().getUserSessionsStream(realm,
-            session.users().getUserByUsername(realm, "user2")).collect(Collectors.toList());
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            assertEquals(0, kcSession.sessions().getUserSessionsStream(realm, kcSession.users().getUserByUsername(realm, "user1"))
+                    .count());
+            List<UserSessionModel> userSessions = kcSession.sessions().getUserSessionsStream(realm,
+                    kcSession.users().getUserByUsername(realm, "user2")).collect(Collectors.toList());
 
-        assertSame(userSessions.size(), 1);
+            assertSame(userSessions.size(), 1);
 
-        for (UserSessionModel userSession : userSessions) {
-            Assert.assertEquals((int) clientSessionsKept.get(userSession.getId()),
-                userSession.getAuthenticatedClientSessions().size());
-        }
+            for (UserSessionModel userSession : userSessions) {
+                Assert.assertEquals((int) clientSessionsKept.get(userSession.getId()),
+                        userSession.getAuthenticatedClientSessions().size());
+            }
+        });
     }
 
     @Test
@@ -327,11 +345,13 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
 
         realm.removeClient(thirdPartyClientUUID);
 
-        for (UserSessionModel s : sessions) {
-            s = session.sessions().getUserSession(realm, s.getId());
-            Set<String> clientUUIDS = s.getAuthenticatedClientSessions().keySet();
-            assertEquals(clientUUIDS, clientSessionsKept.get(s.getId()));
-        }
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            for (UserSessionModel s : sessions) {
+                s = kcSession.sessions().getUserSession(realm, s.getId());
+                Set<String> clientUUIDS = s.getAuthenticatedClientSessions().keySet();
+                assertEquals(clientUUIDS, clientSessionsKept.get(s.getId()));
+            }
+        });
 
         // Revert client
         realm.addClient("third-party");
@@ -376,18 +396,20 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
             // remove the expired sessions - we expect the first two sessions to have been removed as they either expired the max lifespan or the session idle timeouts.
             KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession session1) -> session1.sessions().removeExpired(realm));
 
-            for (String e : expiredUserSessions) {
-                assertNull(session.sessions().getUserSession(realm, e));
-            }
+            KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+                for (String e : expiredUserSessions) {
+                    assertNull(kcSession.sessions().getUserSession(realm, e));
+                }
 
-            for (String v : validUserSessions) {
-                UserSessionModel userSessionLoaded = session.sessions().getUserSession(realm, v);
-                assertNotNull(userSessionLoaded);
-                // the only valid user session should also have a valid client session that hasn't expired.
-                AuthenticatedClientSessionModel clientSessionModel = userSessionLoaded.getAuthenticatedClientSessions().get(client.getId());
-                assertNotNull(clientSessionModel);
-                assertTrue(validClientSessions.contains(clientSessionModel.getId()));
-            }
+                for (String v : validUserSessions) {
+                    UserSessionModel userSessionLoaded = kcSession.sessions().getUserSession(realm, v);
+                    assertNotNull(userSessionLoaded);
+                    // the only valid user session should also have a valid client session that hasn't expired.
+                    AuthenticatedClientSessionModel clientSessionModel = userSessionLoaded.getAuthenticatedClientSessions().get(client.getId());
+                    assertNotNull(clientSessionModel);
+                    assertTrue(validClientSessions.contains(clientSessionModel.getId()));
+                }
+            });
         } finally {
             Time.setOffset(0);
             session.getKeycloakSessionFactory().publish(new ResetTimeOffsetEvent());
@@ -539,12 +561,14 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
     @ModelTest
     public void testGetByClient(KeycloakSession session) {
         RealmModel realm = session.realms().getRealmByName("test");
-        UserSessionModel[] sessions = createSessions(session);
+        final UserSessionModel[] sessions = createSessions(session);
 
-        assertSessions(session.sessions().getUserSessionsStream(realm, realm.getClientByClientId("test-app"))
-                .collect(Collectors.toList()), sessions[0], sessions[1], sessions[2]);
-        assertSessions(session.sessions().getUserSessionsStream(realm, realm.getClientByClientId("third-party"))
-                .collect(Collectors.toList()), sessions[0]);
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession kcSession) -> {
+            assertSessions(kcSession.sessions().getUserSessionsStream(realm, realm.getClientByClientId("test-app"))
+                    .collect(Collectors.toList()), sessions[0], sessions[1], sessions[2]);
+            assertSessions(kcSession.sessions().getUserSessionsStream(realm, realm.getClientByClientId("third-party"))
+                    .collect(Collectors.toList()), sessions[0]);
+        });
     }
 
     @Test
@@ -569,11 +593,13 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
             }
         });
 
-        assertPaginatedSession(session, realm, realm.getClientByClientId("test-app"), 0, 1, 1);
-        assertPaginatedSession(session, realm, realm.getClientByClientId("test-app"), 0, 10, 10);
-        assertPaginatedSession(session, realm, realm.getClientByClientId("test-app"), 10, 10, 10);
-        assertPaginatedSession(session, realm, realm.getClientByClientId("test-app"), 20, 10, 5);
-        assertPaginatedSession(session, realm, realm.getClientByClientId("test-app"), 30, 10, 0);
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession kcSession) -> {
+            assertPaginatedSession(kcSession, realm, realm.getClientByClientId("test-app"), 0, 1, 1);
+            assertPaginatedSession(kcSession, realm, realm.getClientByClientId("test-app"), 0, 10, 10);
+            assertPaginatedSession(kcSession, realm, realm.getClientByClientId("test-app"), 10, 10, 10);
+            assertPaginatedSession(kcSession, realm, realm.getClientByClientId("test-app"), 20, 10, 5);
+            assertPaginatedSession(kcSession, realm, realm.getClientByClientId("test-app"), 30, 10, 0);
+        });
     }
 
     @Test
@@ -676,8 +702,10 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
         RealmModel realm = session.realms().getRealmByName("test");
         createSessions(session);
 
-        assertEquals(3, session.sessions().getActiveUserSessions(realm, realm.getClientByClientId("test-app")));
-        assertEquals(1, session.sessions().getActiveUserSessions(realm, realm.getClientByClientId("third-party")));
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            assertEquals(3, kcSession.sessions().getActiveUserSessions(realm, realm.getClientByClientId("test-app")));
+            assertEquals(1, kcSession.sessions().getActiveUserSessions(realm, realm.getClientByClientId("third-party")));
+        });
     }
 
     @Test
@@ -744,15 +772,20 @@ public class UserSessionProviderTest extends AbstractTestRealmKeycloakTest {
 
         createSessions(session);
 
-        assertEquals(2, session.sessions().getUserSessionsStream(realm, user1).count());
-        assertEquals(1, session.sessions().getUserSessionsStream(realm, user2).count());
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession kcSession) -> {
+            assertEquals(2, kcSession.sessions().getUserSessionsStream(realm, user1).count());
+            assertEquals(1, kcSession.sessions().getUserSessionsStream(realm, user2).count());
+        });
 
-        // remove user1
-        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession kcSession) ->
-                (new UserManager(kcSession)).removeUser(realm, user1));
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession kcSession) -> {
+            // remove user1
+            new UserManager(kcSession).removeUser(realm, user1);
+        });
 
-        assertEquals(0, session.sessions().getUserSessionsStream(realm, user1).count());
-        assertEquals(1, session.sessions().getUserSessionsStream(realm, user2).count());
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession kcSession) -> {
+            assertEquals(0, kcSession.sessions().getUserSessionsStream(realm, user1).count());
+            assertEquals(1, kcSession.sessions().getUserSessionsStream(realm, user2).count());
+        });
     }
 
     private static AuthenticatedClientSessionModel createClientSession(KeycloakSession session, ClientModel client, UserSessionModel userSession, String redirect, String state) {

